vm/src/value.rs

```diff
@@ -202,6 +207,199 @@ impl Value {
     }
 }
 
+#[derive(PartialEq, Copy, Clone, PartialOrd)]
+enum Prec {
+    Top,
+    Constructor,
+}
+use self::Prec::*;
+
+
+pub struct ValuePrinter<'a> {
+    pub typ: &'a ArcType,
+    pub env: &'a TypeEnv,
+    pub value: Value,
+    pub max_level: i32,
+    pub width: usize,
+}
+
+impl<'t> ValuePrinter<'t> {
+    pub fn new(env: &'t TypeEnv, typ: &'t ArcType, value: Value) -> ValuePrinter<'t> {
+        ValuePrinter {
+            typ: typ,
+            env: env,
+            value: value,
+            max_level: i32::max_value(),
+            width: 80,
+        }
+    }
+
+    pub fn max_level(&mut self, max_level: i32) -> &mut ValuePrinter<'t> {
+        self.max_level = max_level;
+        self
+    }
+
+    pub fn width(&mut self, width: usize) -> &mut ValuePrinter<'t> {
+        self.width = width;
+        self
+    }
+}
+
+struct InternalPrinter<'a, 't> {
+    typ: &'t ArcType,
+    env: &'t TypeEnv,
+    arena: &'a Arena<'a>,
+    prec: Prec,
+    level: i32,
+}
+
+impl<'a> fmt::Display for ValuePrinter<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+
+        let arena = Arena::new();
+        let mut s = Vec::new();
+        InternalPrinter {
+                typ: self.typ,
+                env: self.env,
+                arena: &arena,
+                prec: Top,
+                level: self.max_level,
+            }.pretty(self.value)
+            .group()
+            .1
+            .render(self.width, &mut s)
+            .map_err(|_| fmt::Error)?;
+        write!(f, "{}", ::std::str::from_utf8(&s).expect("utf-8"))
+    }
+}
+
+impl<'a, 't> InternalPrinter<'a, 't> {
+    fn pretty(&self, value: Value) -> DocBuilder<'a, Arena<'a>> {
+        use std::iter;
+
+        let arena = self.arena;
+        match value {
+            _ if self.level == 0 => arena.text(".."),
+            Value::String(s) => arena.text(format!("{:?}", s)),
+            Value::Data(ref data) => self.pretty_data(data.tag, data.fields.iter().cloned()),
+            Value::Tag(tag) => self.pretty_data(tag, iter::empty()),
+            Value::Function(ref function) => {
+                chain![arena;
+                    "<extern ",
+                    function.id.declared_name().to_string(),
+                    ">"
+                ]
+            }
+            Value::Closure(ref closure) => {
+                chain![arena;
+                    "{",
+                    arena.concat(closure.upvars.iter().map(|field| {
+                        arena.text(format!("{:?}", field))
+                    })),
+                    "}"
+                ]
```

Comment:
```
arena.text("{")
    .append(arena.concat(closure.upvars.iter().map(|field| {
        arena.text(format!("{:?}", field))
    }))
    .append("}")
```

Just gotten into the habit of using the macro for anything that needs more than a single append. The chain macro also scales better if more stuff are added to the printing (and in this case the debugger PR will actually add some extra information to printing a closure).